core/src/main/java/jenkins/fingerprints/FingerprintStorage.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2020, Sumit Sarin and Jenkins project contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.fingerprints;
+
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+
+import java.io.IOException;
+
+import hudson.model.Fingerprint;
+import jenkins.util.SystemProperties;
+import org.kohsuke.accmod.restrictions.Beta;
+import org.kohsuke.accmod.Restricted;
+
+/**
+ * Pluggable fingerprint storage API for fingerprints.
+ *
+ * @author Sumit Sarin
+ */
+@Restricted(Beta.class)
+public abstract class FingerprintStorage implements ExtensionPoint {
```

Comment:
Do we need to introduce methods which iterate fingerprints without loading them?
AFAICT they are required for effective implementations of  https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/model/FingerprintMap.java . Same might be needed for bulk loading which is a common use-case for Jenkins Web UI

Can be done later from what I can dell

Reply:
I'll investigate this properly, and then comment back

Reply:
@oleg-nenashev I am understanding the concept, but it would be very helpful if you could point me exactly where such implementation has been made, because I was unable to find it.
From what I can understand, `FingerprintMap` is ensuring no two Fingerprint objects for the same ID in memory, and that also memory conservation by making these Fingerprints SoftReference.

Reply:
At any rate, seems that at least `FingerprintMap.isReady` needs to be amended.

Reply:
`FingerprintMap.isReady` has now been updated